File: assign03/packet_analyzer.c
/********************
Lab : Network Lab
Name : Abhinav Kumar
Enroll no.: 510817075
Assignment : 03(Q.01)
*********************/

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <unistd.h>
#include <netinet/in.h>
#include <netinet/if_ether.h>
#include <asm/byteorder.h>
int main()
{
    struct ethhdr
    {
        unsigned char h_dest[ETH_ALEN];
        unsigned char h_source[ETH_ALEN];
        __be16 h_proto;
    } __attribute__((packed));

    int sock_r;
    sock_r = socket(AF_PACKET, SOCK_RAW, htons(ETH_P_ALL));
    if (sock_r < 0)
    {
        printf("error in socket\n");
        return -1;
    }
    printf("Socket creation successful.\n");
    unsigned char *buffer = (unsigned char *)malloc(65536); //to receive data
    memset(buffer, 0, 65536);
    struct sockaddr saddr;
    struct sockaddr_in source, dest;
    int saddr_len = sizeof(saddr);

    //Receive a network packet and copy in to buffer
    int buflen = recvfrom(sock_r, buffer, 65536, 0, &saddr, (socklen_t *)&saddr_len);
    if (buflen < 0)
    {
        printf("error in reading recvfrom function\n");
        return -1;
    }
    struct ethhdr *eth = (struct ethhdr *)(buffer);
    printf("\nEthernet Header\n");
    printf("\t | -Source Address: % .2X - % .2X - % .2X - % .2X - % .2X - % .2X\n", eth->h_source[0], eth->h_source[1], eth->h_source[2], eth->h_source[3], eth->h_source[4], eth->h_source[5]);
    printf("\t | -Destination Address: % .2X - % .2X - % .2X - % .2X - % .2X - % .2X\n", eth->h_dest[0], eth->h_dest[1], eth->h_dest[2], eth->h_dest[3], eth->h_dest[4], eth->h_dest[5]);
    printf("\t | -Protocol: % d\n", eth->h_proto);
}

File: assign03/packet_analyzer.py
##########################
# Lab : Network Lab      #
# Name : Abhinav Kumar   #
# Enroll no.: 510817075  #
# Assignment : 03(Q.01)  #
##########################


# works on unix/linux
import socket
import struct
import binascii


class PacketAna():
    def __init__(self, raw_socket):
        self.raw_socket = raw_socket
        self.packet = self.raw_socket.recvfrom(2048)

    def ethr_ana(self):
        # ether header
        ethrow_header = self.packet[0][0:14]
        # 6destMac6hostMac2etherType
        ether_header = struct.unpack("!6s6s2s", ethrow_header)
        print(f"\t |-Source address = {binascii.hexlify(ether_header[0])}")
        print(
            f"\t |-Destination address = {binascii.hexlify(ether_header[1])}")
        print(f"\t |-Protocol = {binascii.hexlify(ether_header[2])}")

    def ip_ana(self):
        ipro_header = self.packet[0][14:34]
        ip_header = struct.unpack("!12s4s4s", ipro_header)
        # networktoascii
        print(f"\t |-Source IP is = {socket.inet_ntoa(ip_header[1])}")
        print(f"\t |-Destination IP = {socket.inet_ntoa(ip_header[2])}")

    def tcp_ana(self):
        tcpro_header = self.packet[0][34:54]
        tcp_header = struct.unpack("!HH16s", tcpro_header)
        print(f"\t |-Source port = {tcp_header[0]}")
        print(f"\t |-Destination port = {tcp_header[1]}")


if __name__ == "__main__":
    raw_socket = socket.socket(
        family=socket.PF_PACKET, type=socket.SOCK_RAW, proto=socket.htons(0x0800))
    ana = PacketAna(raw_socket)
    print("\t***********PACKET**FROM**RAWSOCKET***********")
    print(f"\t |-({ana.packet})")
    print("\n\t***********FROM**ETH**HEADER**ANALYSIS************")
    ana.ethr_ana()
    print("\n\t***********FROM**IP**HEADER**ANALYSIS**********")
    ana.ip_ana()
    print("\n\t***********FROM**TCP**HEADER**ANALYSIS************\n")
    ana.tcp_ana()
